ember-server/src/test/scala/org/http4s/ember/server/EmberServerMtlsSuite.scala

```diff
@@ -0,0 +1,163 @@
+
+import cats.effect._
+import fs2.io.net.tls.TLSContext
+import fs2.io.net.tls.TLSParameters
+import java.net.URL
+import java.nio.charset.StandardCharsets
+import java.security.KeyStore
+import javax.net.ssl._
+import org.http4s._
+import org.http4s.server.Server
+import org.http4s.implicits._
+import org.http4s.dsl.Http4sDsl
+import org.http4s.server.ServerRequestKeys
+import org.http4s.testing.ErrorReporting
+import scala.io.Source
+import scala.util.Try
+
+/** Test cases for mTLS support in Ember Server.
+  */
+class EmberServerMtlsSuite extends Http4sSuite {
+  {
+    val hostnameVerifier: HostnameVerifier = new HostnameVerifier {
+      override def verify(s: String, sslSession: SSLSession): Boolean = true
+    }
+
+    // For test cases, don't do any host name verification. Certificates are self-signed and not available to all hosts
+    HttpsURLConnection.setDefaultHostnameVerifier(hostnameVerifier)
+  }
+
+  def service[F[_]](implicit F: Async[F]): HttpApp[F] = {
+    val dsl = new Http4sDsl[F] {}
+    import dsl._
+
+    HttpRoutes
+      .of[F] {
+        case req @ GET -> Root / "dummy" =>
+          val output = req.attributes
+            .lookup(ServerRequestKeys.SecureSession)
+            .flatten
+            .map { session =>
+              assert(session.sslSessionId != "")
+              assert(session.cipherSuite != "")
+              assert(session.keySize != 0)
+
+              session.X509Certificate.head.getSubjectX500Principal.getName
+            }
+            .getOrElse("Invalid")
+          Ok(output)
+        case req @ GET -> Root / "noauth" =>
+          req.attributes
+            .lookup(ServerRequestKeys.SecureSession)
+            .flatten
+            .foreach { session =>
+              assert(session.sslSessionId != "")
+              assert(session.cipherSuite != "")
+              assert(session.keySize != 0)
+              assert(session.X509Certificate.isEmpty)
+            }
+
+          Ok("success")
+      }
+      .orNotFound
+  }
+
+  lazy val sslContext: SSLContext = {
+    val ks = KeyStore.getInstance("JKS")
+    ks.load(getClass.getResourceAsStream("/keystore.jks"), "password".toCharArray)
+
+    val kmf = KeyManagerFactory.getInstance("SunX509")
+    kmf.init(ks, "password".toCharArray)
+
+    val js = KeyStore.getInstance("JKS")
+    js.load(getClass.getResourceAsStream("/keystore.jks"), "password".toCharArray)
+
+    val tmf = TrustManagerFactory.getInstance("SunX509")
+    tmf.init(js)
+
+    val sc = SSLContext.getInstance("TLSv1.2")
+    sc.init(kmf.getKeyManagers, tmf.getTrustManagers, null)
+
+    sc
+  }
+
+  lazy val noAuthClientContext: SSLContext = {
+    val js = KeyStore.getInstance("JKS")
+    js.load(getClass.getResourceAsStream("/keystore.jks"), "password".toCharArray)
+
+    val tmf = TrustManagerFactory.getInstance("SunX509")
+    tmf.init(js)
+
+    val sc = SSLContext.getInstance("TLSv1.2")
+    sc.init(null, tmf.getTrustManagers, null)
+
+    sc
+  }
```

Comment:
I think you can build the `TLSContext` directly via `Network[IO]`. See fs2: https://github.com/typelevel/fs2/blob/main/io/jvm/src/test/scala/fs2/io/net/tls/TLSSuite.scala

Reply:
This change requires the tests to be rewritten, because of the use of this call in get:
`conn.setSSLSocketFactory(sslContext.getSocketFactory)` (i.e. we need the SSL Context factory). 
However, we can't redefine `noAuthClientContext` to use `Network[IO]` because there's no way to independently
specify the keystore and truststore when using `Network[IO]` (same keystore is used for both). As a side note, in our 
work projects it's normal to have separate keystore and truststore files when creating a TLS context, so I think
this is an area of potential improvement for Network[IO]. (I'd be happy to submit a PR for that eventually).

Regardless, I did go down the path of rewriting the tests and hit what I think is another issue. 
In the original test code, we could specify the protocol version (i.e. TLSv1.2), but we can't when using `Network[IO]`. This resulted in TLSv1.3 being used, and this exposes an issue with `org.http4s.internal.tls deduceKeyLength`. It doesn't support the TLSv1.3 cipher suites it looks like. So we get a 0 for the key length.

```
@@@ 147ef8d27ed1a3c37e7ef734b106fd41556f3f21fe9082d5a5eb6d89009b3528
@@@ TLS_AES_128_GCM_SHA256
@@@ 0
```

Notice the cipher suite is TLS_AES_128_GCM_SHA256, but that's not handled explicitly in this code resulting in 0:
```
  def deduceKeyLength(cipherSuite: String): Int =
    if (cipherSuite == null) 0
    else if (cipherSuite.contains("WITH_AES_256_")) 256
    else if (cipherSuite.contains("WITH_RC4_128_")) 128
    else if (cipherSuite.contains("WITH_AES_128_")) 128
    else if (cipherSuite.contains("WITH_RC4_40_")) 40
    else if (cipherSuite.contains("WITH_3DES_EDE_CBC_")) 168
    else if (cipherSuite.contains("WITH_IDEA_CBC_")) 128
    else if (cipherSuite.contains("WITH_RC2_CBC_40_")) 40
    else if (cipherSuite.contains("WITH_DES40_CBC_")) 40
    else if (cipherSuite.contains("WITH_DES_CBC_")) 56
    else 0
}
```
I think it should be updated to include TLSv1.3 ciphers. I could write up an issue for that as well. 

So getting back to the review, I'd suggest going with the current implementation for now, and then I can
update the tests at a later date, after the deduceKeyLength issue has been addressed, or we're able to
specify the TLS version via a new version of Network[IO].fromKeystoreResource.

Reply:
Hmm, thanks for the deep dive, that uncovered quite a bit. Yes please, definitely at least raise an issue in fs2 about this if you don't mind (pinging @mpilquist).

> I'd suggest going with the current implementation for now, and then I can update the tests at a later date

To give a little more context, the reason I'm trying so hard to avoid this is because I'm thinking ahead to when this PR gets merged into `main` where Ember now runs on both JVM and Node.js (Javascript via Scala.js). So the more of this test that can be written without referencing Java classes which are not available on Scala.js the easier that merge will be. Having a completely cross-platform test is unlikely though, but ideally we could share more code. Or maybe not, in fs2 the TLS tests are implemented twice for JVM and Node.js each which is unfortunate.

Reply:
I think using `fromSSLContext` is right approach when there's a separate keystone and Truststore.

Re: protocols, use `TLSParameters#protocols` like here: https://github.com/typelevel/fs2/blob/d0b358c165e575474575dd871f8dfafafac204ee/io/jvm/src/test/scala/fs2/io/net/tls/TLSSocketSuite.scala#L47-L54

Reply:
If you'll pardon my ignorance, why exactly is a seperate truststore necessary for this test?

Reply:
For the test `Server should handle mTLS request correctly without clientAuth`, we test `server = want client auth`, but the client's TLS context is created without a keystore (but with a truststore), so there's no key/cert to send. 

Update regarding rewriting the tests: I was able to rewrite one of the tests to use Network[IO].fromKeystoreResource and get it to work by passing a TLSParameters instance with the protocol set to TLSv1.2 to EmberServerBuilder. If I do this for the rest of them I can get rid of the hostnameVerifier code mentioned in the other comment. There'd still be the need to use `fromSSLContext` for the no client auth case mentioned above, but the tests would be closer to what you're looking for I think. 

One thing to point out about EmberClientBuilder in the 0.23 branch: It doesn't provide the ability to pass in TLSParameters, which means I don't think you could do what Mike was demonstrating with TLSContext client
Builder with the current EmberClientBuilder. It looks like this has been added in the main 1.0.0... branch though.